File: src/android/CustomTestHandler.kt
package com.ncinga.speedtest

import android.util.Log
import com.ookla.speedtest.sdk.*
import com.ookla.speedtest.sdk.config.Config
import com.ookla.speedtest.sdk.config.ValidatedConfig
import com.ookla.speedtest.sdk.handler.TaskManagerController
import com.ookla.speedtest.sdk.handler.TestHandlerBase
import com.ookla.speedtest.sdk.model.LatencyResult
import com.ookla.speedtest.sdk.model.TransferResult
import com.ookla.speedtest.sdk.result.OoklaError
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.apache.cordova.CallbackContext
import org.apache.cordova.PluginResult
import org.json.JSONObject

class CustomTestHandler(
    private val speedtestSDK: SpeedtestSDK,
    private val configName: String,
    private val timeInterval: Int,
    private val endpoint: String,
    private val callbackContext: CallbackContext

) : TestHandlerBase() {
    private var taskManager: TaskManager? = null
    private val TAG = "SpeedTest"
    private var endpointHandler: EndpointHandler? = null
    fun runTestWithSingleServer(
        count: Int,
        currentIteration: Int = 1
    ) {
        endpointHandler = EndpointHandler(endpoint)
        if (count != -1 && currentIteration > count) {
            Log.i(TAG, "All Test completed")
            return
        }

        Log.i(TAG, "Iteration $currentIteration")
        val config = Config.newConfig(configName)

        val configHandler = object : ConfigHandlerBase() {
            override fun onConfigFetchFinished(validatedConfig: ValidatedConfig?) {
                val handler = object : TestHandlerBase() {
                    override fun onLatencyFinished(
                        taskController: TaskManagerController?,
                        result: LatencyResult
                    ) {
                        super.onLatencyFinished(taskController, result)
                        Log.d(TAG, "Latency Result: $result")
                        taskManager?.startNextStage()
                    }

                    override fun onUploadFinished(
                        taskController: TaskManagerController?,
                        result: TransferResult
                    ) {
                        super.onUploadFinished(taskController, result)
                        Log.d(TAG, "Upload Speed: $result")
                        taskManager?.startNextStage()
                    }

                    override fun onDownloadFinished(
                        taskController: TaskManagerController?,
                        result: TransferResult
                    ) {
                        super.onDownloadFinished(taskController, result)
                        Log.d(TAG, "Download Speed: $result")
                        taskManager?.startNextStage()
                    }

                    override fun onTestFinished(speedtestResult: SpeedtestResult) {
                        super.onTestFinished(speedtestResult)
                        val result = speedtestResult.getResult().toJsonString()
                        sendUpdate(result, currentIteration)

                        CoroutineScope(Dispatchers.Main).launch {
                            delay(timeInterval.toLong() * 1000)
                            runTestWithSingleServer(count, currentIteration + 1)
                        }
                    }

                    override fun onTestFailed(
                        error: OoklaError,
                        speedtestResult: SpeedtestResult?
                    ) {
                        super.onTestFailed(error, speedtestResult)
                        Log.e(TAG, error.message)
                        sendError(error.message ?: "Unknown error")

                        CoroutineScope(Dispatchers.Main).launch {
                            delay(timeInterval.toLong() * 1000)
                            runTestWithSingleServer(count, currentIteration + 1)
                        }
                    }
                }

                taskManager = speedtestSDK.newTaskManager(handler, validatedConfig)
                taskManager?.start()
            }

            override fun onConfigFetchFailed(error: OoklaError) {
                Log.e(TAG, "Config fetch failed with: ${error.message}")
                sendError("Config fetch failed: ${error.message}")
            }
        }

        ValidatedConfig.validate(config, MainThreadConfigHandler(configHandler))
    }

    private fun sendUpdate(result: String, iteration: Int) {
        val jsonObject = JSONObject()
        jsonObject.put("iteration", iteration)
        val jsonResult = JSONObject(result)
        jsonObject.put("result", jsonResult)
        val pluginResult = PluginResult(PluginResult.Status.OK, jsonObject)
        pluginResult.keepCallback = true
        endpointHandler?.sendData(jsonObject.toString())
        callbackContext.sendPluginResult(pluginResult)
    }

    private fun sendError(message: String) {
        val pluginResult = PluginResult(PluginResult.Status.ERROR, message)
        pluginResult.keepCallback = false
        callbackContext.sendPluginResult(pluginResult)
    }

    fun stopTesting() {
        val result = JSONObject();
        result.put("message", "stop testing")
        taskManager?.cancel()
        val pluginResult = PluginResult(PluginResult.Status.OK, result)
        pluginResult.keepCallback = true
        callbackContext.sendPluginResult(pluginResult)
    }
}

File: src/ios/CustomTestHandler.swift
import Foundation
import SpeedtestSDK
import CoreLocation
import os

public class CustomTestHandler {
    private var testResult: [ResultDTO] = []
    private var speedTestsdk: STSpeedtestSDK?
    private var configName: String?
    private var taskManager: STTaskManager?
    private var callback: (([String: Any]) -> Void)?
    private var count: Int32
    private var endpointHandler: EndpointHandler?
    private var timeInterval: Int32?

    init(speedTestsdk: STSpeedtestSDK?, configName: String?, count: Int32 = 1,
         endpoint: String?, timeInterval: Int32?, callback: (([String: Any]) -> Void)?) {
        self.speedTestsdk = speedTestsdk
        self.configName = configName
        self.count = count
        self.callback = callback
        self.timeInterval = timeInterval
        self.endpointHandler = EndpointHandler(endpoint: endpoint ?? "")
    }

    func runTestWithSingleServer() {
        guard count > 0 else {
            os_log("Invalid count. Cannot run test.")
            return
        }

        runTestIteration(currentIteration: 1)
    }

    private func runTestIteration(currentIteration: Int) {
        guard currentIteration <= count else {
            finalizeResults()
            return
        }

        os_log("Running test iteration %d of %d", currentIteration, count)

        let config = STConfig.newConfig(configName ?? "SpeedTest")
        config?.validate { [weak self] validatedConfig, error in
            guard let self = self, let validatedConfig = validatedConfig else {
                os_log("Failed to load config: %@", error?.message ?? "Unknown Error")
                return
            }

            let handler = TestHandler(currentIteration: currentIteration, testResult: &self.testResult) {
                let delay = Double(self.timeInterval ?? 10)
                DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
                    self.runTestIteration(currentIteration: currentIteration + 1)
                }
            }

            do {
                self.taskManager = try self.speedTestsdk?.newTaskManager(STMainThreadTestHandler(delegate: handler), config: validatedConfig)
                handler.taskManager = self.taskManager
                self.taskManager?.start()
            } catch {
                os_log("Failed to create task manager: %@", "\(error)")
            }
        }
    }

    private func finalizeResults() {
        os_log("All tests completed.")
        do {
            guard let testResults = testResult as? [ResultDTO] else {
                os_log("testResult is not of type [ResultDTO].")
                return
            }
            let resultArray = convertTestResultsToJsonArray(testResults: testResults)
            self.endpointHandler?.sendData(body: convertTestResultsToJson(testResults: testResults))
            callback?(resultArray)
        } catch {
            os_log("Error processing result: \(error.localizedDescription)")
        }
    }

    class TestHandler: NSObject, STTestHandlerDelegate, ObservableObject {
        weak var taskManager: STTaskManager?
        private var currentIteration: Int
        private var completionHandler: () -> Void
        private var testResult: UnsafeMutablePointer<[ResultDTO]>

        init(currentIteration: Int, testResult: inout [ResultDTO], completionHandler: @escaping () -> Void) {
            self.currentIteration = currentIteration
            self.testResult = UnsafeMutablePointer(&testResult)
            self.completionHandler = completionHandler
        }

        func onResultUploadFinished(result: SpeedtestSDK.STResult.ResultUpload?, _ error: SpeedtestSDK.STResult.OoklaError?) {
            taskManager?.startNextStage()
        }

        func onUploadFinished(_ taskController: STTaskManagerController?, result: STTransferResult) {
            taskManager?.startNextStage()
        }

        func onLatencyFinished(_ taskController: STTaskManagerController?, result: STLatencyResult) {
            taskManager?.startNextStage()
        }

        func onDownloadFinished(_ taskController: STTaskManagerController?, result: STTransferResult) {
            taskManager?.startNextStage()
        }

        func onTestFinished(_ result: SpeedtestSDK.STSpeedtestResult) {
            do {
                let jsonData = try JSONEncoder().encode(result.getResult())
                if let jsonString = String(data: jsonData, encoding: .utf8) {
                    let loopResult = ResultDTO(id: "loop\(currentIteration)", result: jsonString)
                    testResult.pointee.append(loopResult)
                } else {
                    os_log("Error converting JSON data to string.")
                }
            } catch {
                os_log("Failed to get JSON result: %@", error.localizedDescription)
            }
            taskManager?.startNextStage()
            completionHandler()
        }
    }
}
